### DISCUSSION TITLE
[javascript deep dive] 6장. 데이터 타입

### DISCUSSION THREAD

**Discussion by rladpwl0512** [woowacourse-study/2022-javascript-bank#5] [General]
[javascript deep dive] 6장. 데이터 타입

**Comment by jin7969** [comment]
데이터 타입이 왜 필요할까?

**Reply by rladpwl0512** [reply]
자바스크립트 엔진은 타입을 구별해서 값을 취급함으로써 타입에 맞는 메모리 공간을 확보하고, 들어온 데이터를 해석하는 방식을 정한다.

**Reply by jin7969** [reply]
왜 숫자는 64비트로 고정이고 문자는 길이마다 달라지는가??

**Comment by jswith** [comment]
자바스크립트의 데이터타입에 대해 설명해주세요

**Reply by jswith** [reply]
- 원시 타입, 객체 타입
- 차이점은 원시는 불변, 객체는 가변
- 불변 : 뭐가 변하지 않는가? 값이 변하지 않는다. 어디서 변하지 않을까? 메모리 내의 값
- 가변 : 뭔가 변한다. 뭐가 변할까? 값이 변한다. 어디서 변할까? 객체의 변수가 가리키는 메모리 셀이 가리키는 객체의 메모리 셀

**Comment by rladpwl0512** [comment]
undefined, null의 공통점과 차이점에 대해 설명하고, 각각을 언제 사용할까요?

**Reply by rladpwl0512** [reply]
`undefined`, `null`은 두개 모두 값이 없음을 나타낸다. 
- `undefined`: 시스템상에서 값이 없을 때 
- `null`: 개발자가 명시적으로 값이 없음을 나타낼 때 / 함수가 유효한 값을 반환할 수 없는 경우 (ex-`document.querySelector`는 조건에 부합하는 html 요소를 검색할 수 있다.)

**Comment by rladpwl0512** [comment]
- 동적타입언어, 정적타입언어에 대해 설명해주세요. 
- Javascript는 어디에 속하나요? 
  (Javascript 변수의 타입이 결정되는 방식을 설명해주세요.)

**Reply by rladpwl0512** [reply]
- 동적 타입언어: 할당시기에 타입 결정 (아무 데이터 타입의 값을 할당할 수 있기 때문에 위험하다) -ex. javascript 
- 정적 타입언어: 선언시기에 타입 결정 (미리 선언시기에 작성한 타입의 값만 할당할 수 있다) -ex. typescript (js의 위험성을 개선)
  - 그럼 정적 타입언어는 무조건 좋은가요? 개발자가 힘들다, 가벼운 프로젝트에서(타입이 모두 예상이 가능할 때)는 괜히 썼다가 프로그램에 부담이 될 수 있다.
